inc/AbstractPdoDbDriver.php

```diff
@@ -0,0 +1,713 @@
+			{
+				$flags = array(
+					PDO::ATTR_PERSISTENT => false,
+					PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION,
+					PDO::ATTR_EMULATE_PREPARES => false,
+				);
+
+				if (!empty($singleConnection['pconnect'])) {
+					$flags[PDO::ATTR_PERSISTENT] = true;
+				}
+
+				$link = "{$type}_link";
+
+				get_execution_time();
+
+				list($hostname, $port) = self::parseHostname($singleConnection['hostname']);
+
+				$dsn = $this->getDsn(
+					$hostname,
+					$config['database'],
+					$port,
+					$this->db_encoding
+				);
+
+				try {
+					$this->$link = new PDO(
+						$dsn,
+						$singleConnection['username'],
+						$singleConnection['password'],
+						$flags
+					);
+
+					$this->lastPdoException = null;
+				} catch (PDOException $e) {
+					$this->$link = null;
+					$this->lastPdoException = $e;
+				}
+
+				$time_spent = get_execution_time();
+				$this->query_time += $time_spent;
+
+				// Successful connection? break down brother!
+				if ($this->$link !== null) {
+					$this->connections[] = "[".strtoupper($type)."] {$singleConnection['username']}@{$singleConnection['hostname']} (Connected in ".format_time_duration($time_spent).")";
+					break;
+				} else {
+					$this->connections[] = "<span style=\"color: red\">[FAILED] [".strtoupper($type)."] {$singleConnection['username']}@{$singleConnection['hostname']}</span>";
+				}
+			}
+		}
+
+		// No write server was specified (simple connection or just multiple servers) - mirror write link
+		if (empty($connections['write'])) {
+			$this->write_link = $this->read_link;
+		}
+
+		// Have no read connection?
+		if ($this->read_link === null) {
+			$this->error("[READ] Unable to connect to database server");
+			return false;
+		} else if($this->write_link === null) {
+			$this->error("[WRITE] Unable to connect to database server");
+			return false;
+		}
+
+		$this->database = $config['database'];
+
+		if (version_compare('PHP_VERSION', '5.3.6', '<') === true) {
+			// character set in DSN was ignored before PHP 5.3.6, so we must SET NAMES
+			$this->setCharacterSet($this->db_encoding);
+		}
+
+		$this->current_link = $this->read_link;
+		return true;
+	}
+
+	/**
+	 * Parse a hostname and possible port combination.
+	 *
+	 * @param string $hostname The hostname string. Can be any of the following formats:
+	 * - `127.0.0.1` - IPv4 address.
+	 * - `[::1]` - IPv6 address.
+	 * - `localhost` - hostname.
+	 * - `127.0.0.1:3306` - IPv4 address and port combination.
+	 *  `[::1]:3306` - IPv6 address and port combination.
+	 * - `localhost:3306` - hostname and port combination.
+	 *
+	 * @return array Array of host and port.
+	 *
+	 * @throws InvalidArgumentException Thrown if {@see $hostname} is an IPv6 address which lacks a closing square bracket.
+	 */
+	private static function parseHostname($hostname)
+	{
+		// first, check for an IPv6 address - IPv6 addresses always start with `[`
+		$openingSquareBracket = strpos($hostname, '[');
+		if ($openingSquareBracket === 0) {
```

Comment:
Note: This logic is not yet tested. I believe that other drivers doing much the same thing would never have worked with IPv6 addressed DB hosts...